lib/pundit.rb

```diff
@@ -61,17 +61,23 @@ class << self
     # authorized to perform the given action.
     #
     # @param user [Object] the user that initiated the action
-    # @param record [Object] the object we're checking permissions of
+    # @param possibly_namespaced_record [Object, Array] the object we're checking permissions of
     # @param query [Symbol, String] the predicate method to check on the policy (e.g. `:show?`)
     # @param policy_class [Class] the policy class we want to force use of
+    # @param cache [#[], #[]=] a Hash-like object to cache the found policy instance in
     # @raise [NotAuthorizedError] if the given query method returned false
     # @return [Object] Always returns the passed object record
-    def authorize(user, record, query, policy_class: nil)
-      policy = policy_class ? policy_class.new(user, record) : policy!(user, record)
+    def authorize(user, possibly_namespaced_record, query, policy_class: nil, cache: {})
+      record = pundit_model(possibly_namespaced_record)
+      policy = if policy_class
+        policy_class.new(user, record)
+      else
+        cache[possibly_namespaced_record] ||= policy!(user, possibly_namespaced_record)
```

Comment:
This appears to be a breaking change, and I'm curious whether or not that was intentional. Our code uses the `authorize(record)` method below which [used to find the policy_class with `policy`](https://github.com/varvet/pundit/pull/697/files#diff-2acc48fe8302d7198f5a4aa4aaa8b4f0502d185e186a099283264da056d450c7L221) but this now uses `policy!` in that situation. 

Can you please verify that you meant to force a more explicit definition of `policy_class`? Thanks!

Reply:
Knee-jerk reaction is that's indeed unintended! We had some other things that were affected by this move that might also be unintended, so I'd wager it's very likely we'll make some adjustments to this code in the near future.

Reply:
Thanks for the reply and opening a new issue. I'll stick to 2.1.1 until this is worked out. Let me know if there's anything I can do to help.

Reply:
Hi @adherr! This was a long time ago, but I'm here again for other reasons.

While the previous implementation used to use `policy(record)`, it still relied on the method returning an actual policy because the next call would be `policy.public_send(query)` which would fail with a `NoMethodError` if `nil` was returned for the policy.

The main change is that after this we instead raise `Pundit::NotDefinedError` as opposed to the `NoMethodError`.

What part of this behaviour were you relying on that prevented the upgrade?